src/main/java/com/puppycrawl/tools/checkstyle/CheckCloneService.java

```diff
@@ -0,0 +1,156 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2017 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.stream.Collectors;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.CheckstyleException;
+import com.puppycrawl.tools.checkstyle.api.Configurable;
+import com.puppycrawl.tools.checkstyle.api.Configuration;
+import com.puppycrawl.tools.checkstyle.api.Contextualizable;
+import com.puppycrawl.tools.checkstyle.api.FileSetCheck;
+
+/**
+ * A service for clonning check instances.
+ * @author Andrew Kuchev.
+ */
+public final class CheckCloneService {
+    /** Logger for CheckCloneService. */
+    private static final Log LOG = LogFactory.getLog(CheckCloneService.class);
+
+    /** A log pattern for case when a check does not implement appropriate interface. */
+    private static final String CHECK_MUST_IMPLEMENT_INTERFACE_MSG =
+            "Check %s should implement either OneCheckInstancePerThread "
+                    + "or OneCheckInstancePerApplication%n";
+
+    /** A private constructor is required for utility class. */
+    private CheckCloneService() {
+    }
+
+    /**
+     * Clones the given check.
+     * @param check A check to be cloned.
+     * @return Cloned check.
+     * @noinspection InstanceofChain
+     */
+    // we cannot move this logic to the AbstractCheck subclasses,
+    // because they should not know about PackageObjectFactory
+    // and other configuration details
```

Comment:
@soon It is not clear at all that comments are for the `noinspection`.
Looking at other `noinspections` I don't think we ever documented them, unlike our own Checkstyle violations and I am not sure if it is allowed.
I would just remove the comments.